Add support for pinned and locked tabs to wxAUI

vadz opened the pull request:
This implements support for the usual concept of pinned tabs (shown before all the other normal tabs) as well as the locked tabs from #25072 (shown even before the pinned tabs and can't be closed nor drag-moved by the user, even among the tabs of the given group).

After reviewing handling of the pinned tabs in different programs, I've discovered that there are quite a few different ways in which they're handled:

- Some programs don't show any special button for them at all, only allow [un]pinning them from a menu.
- Some programs show the pin/unpin button only on the current tab.
- Some programs show the pin button only on the current tab, but unpin button on all pinned tabs.

And this is why we have `wxAUI_NB_PIN_ON_ACTIVE_TAB` and `wxAUI_NB_UNPIN_ON_ALL_PINNED` styles that provide support for all of the possibilities above. Note that I couldn't find any program that always showed pin button on all tabs and experimenting with it showed why: this just makes the display too visually busy. OTOH MSVS shows pin/unpin buttons on all tabs, but only when the mouse is hovering over it and this looks quite nice (but wastes space on the not shown buttons) and is something we could implement later.

As always, any comments/remarks/suggestions are welcome!

vadz committed:
Reuse GetCloseButtonState() logic in wxAuiTabContainer::Render()

Don't duplicate it to avoid it getting out of sync in the two places.

No real changes yet.

vadz committed:
Factor out MovePage(oldIdx, newIdx) overload

No real changes yet, just make the function for moving the page from one
index to another directly available to wxAuiNotebook, which will need it
soon.

vadz committed:
Fix code for computing total wxAuiTabContainer buttons width

The code added in 770e0bcd16 (Improve high DPI support in wxAui,
2018-09-14) was strange: it computed the sum of button widths by
iterating over the buttons in reverse order and, worse, reused the same
loop variable that was also used in the containing loop, changing it and
making the loop exit after handling wxAUI_BUTTON_RIGHT.

This didn't really matter until now because it was the last button
handled in this loop in any case, but fix it before adding code to
handle more buttons -- and also because the new version is simpler and
more obvious.

vadz committed:
More minor fixes to wxAuiTabContainer button state code

Fix the comment and use "else if" for mutually exclusive "if"s.

No real changes.

vadz committed:
Add incomplete support for locked and pinned tabs

Locked tabs can't be closed by the user, while pinned tabs can be closed
but can't be repositioned and are always shown before the normal tabs
(but after the locked ones, if any).

There is no support for interactively [un]pinning tabs yet nor even any
indication that a tab is pinned, but this will be added soon.

vadz committed:
Fix drawing tab "Close" button in hot/pressed states

This was broken, apparently quite some time ago, and the button was
always drawn in normal state as we lost the temporary HOVER or PRESS
state flags when determining whether it should be drawn at all.

vadz committed:
Update GetTabSize() documentation to use wxReadOnlyDC

This function doesn't take wxDC any more.

vadz committed:
Change GetCloseButtonState() to return wxAuiPaneButtonState

This is a bit more type-safe than returning just an int.

No real changes.

vadz committed:
Make wxAuiTabContainer::m_buttons vector instead of legacy array

No real changes, just stop using wxAuiTabContainerButtonArray for this
member, it's protected, so it's formally a backwards-incompatible
change, but it was never documented and people are really not supposed
to inherit from wxAuiTabContainer anyhow, so hopefully nobody should
have been using it.

vadz committed:
Add constructors to wxAuiTabContainerButton

Simplify the code a bit by allowing these objects to be constructed from
initializer lists.

No real changes.

vadz committed:
Declare wxAuiTabContainerButton before wxAuiNotebookPage

Just move the class declaration to allow using it in wxAuiNotebookPage
in the upcoming commit.

No real changes, use Git --color-moved option to see it.

vadz committed:
Implement support for pinned tabs in wxAuiNotebook

Add wxAUI_NB_TAB_PIN style and show "pin"/"unpin" buttons for the tabs
when it is used, allowing to change their kind.

This required adding the page buttons to wxAuiNotebookPage itself, as
passing the data about both buttons (and any other ones that might get
added in the future) to wxAuiTabArt would be too awkward, while with the
buttons available in wxAuiNotebookPage, it is enough to just extend
wxAuiTabArt with the new DrawPageTab() and GetPageTabSize() functions
that are even simpler than the existing DrawTab() and GetTabSize() ones.

Implement the new functions in wxAuiGenericTabArt to handle pin buttons.

Also extend the sample to show the current tab state when dumping all
tabs.

vadz committed:
Restore MSW look of wxAuiNotebook when not using pinned tabs

Due to the change in wxAuiNotebook code, overridden DrawTab() was not
called any longer and DrawPageTab() inherited from the base class was
used instead. Override DrawPageTab(), and also GetPageTabSize(), in
wxAuiMSWTabArt too to restore the old behaviour.

Also, because this implementation doesn't support pinned tabs (yet),
don't use it when wxAUI_NB_TAB_PIN is on.

vadz committed:
Document incompatible change to wxAuiGenericTabArt as base class

While we keep compatibility with the class inheriting from wxAuiTabArt
by redirecting DrawPageTab() to their existing DrawTab(), we can't do
this for the classes deriving from wxAuiGenericTabArt because
DrawPageTab() is overridden in this class and so will be used.

vadz committed:
Allow dragging pinned tabs over other pinned tabs

Dragging pinned tabs shouldn't be entirely forbidden, unlike for the
locked ones, so do allow it and just restrict them from being dragged to
a group of tabs of different kind.

vadz committed:
Fix computing multi-row layout in presence of locked tabs

We can't force the page to take its "Close" button into account when
computing its size if it is locked, because the button is never shown in
this case. And, moreover, we shouldn't do it, because we could add extra
space unnecessarily if the row consists entirely of locked tabs.

So don't adjust the close button state for the locked pages and only do
it when we find the first page of another kind in LayoutMultiLineTabs().

vadz committed:
Replace wxAUI_NB_TAB_PIN with wxAUI_NB_PIN_ON_ACTIVE_TAB

Showing "pin" icon on all tabs is too visually noisy, so (optionally)
show it only on the active tab instead.

Also change SetPageKind() to allow pinning tabs even if no special
styles are used, we can trust the program to know what it is doing and
it is common to use a tab context menu for [un]pinning the tabs instead
of using the built-in button for this, which is allowed now.

vadz committed:
Add wxAUI_NB_UNPIN_ON_ALL_PINNED style

Allow showing "unpin" button on all pinned tabs to visually indicate
that they are pinned and, also, to allow unpinning them without
selecting them first.

vadz committed:
Serialize wxAuiNotebook page kinds too

Save and restore the locked and pinned pages, if any.

Update the serializer in the sample to handle the new wxAuiTabLayoutInfo
fields.

vadz committed:
Show pin/unpin tabs in the AUI sample by default

Enable the new styles for wxAuiNotebook used in the sample to be able to
see how they work by default.

vadz committed:
Improve default wxAuiNotebook theme documentation

Mention MSW-specific theme and document when it is [not] used.

craftyjon commented:
I did a simple test of this branch and it works well for what I need!  Thank you.

vadz committed:
Forbid dragging tabs in invalid position in another tab control

We already checked that dragging tabs inside the same tab control was
only allowed when they remained among the tabs of the same kind (e.g.
pinned or normal), but the same check was missing when dragging to a
different tab control, allowing to drag a normal tab into the group of,
or even before, the pinned tabs.

Check for this too now to prevent this from happening as well.

vadz committed:
Don't draw misleading drop hint if tab can't be dropped

Check that dropping the tab at the current point would be allowed and
hide the drop hint if it isn't, e.g. because we're dragging a pinned tab
which can't be dropped after a normal tab in another tab control.
